### DISCUSSION TITLE
Adequate team size and financial commitment from Microsoft to fix the >2500 bugs (& growing) in MAUI (or at least get the most basic core functions like Border/Editor/Image/Layout working)

### DISCUSSION THREAD

**Discussion by jonmdev** [dotnet/maui#17917] [General]
### Description

**Advantages of MAUI**

I am new to MAUI (only been here 3 weeks) but I already think it is a fantastic system. I was previously working in systems that are completely disconnected from the underlying OS. The advantages of having a system like MAUI where you are directly interfacing with the underlying OS to use native elements are immense and don't need to be repeated. That is why we are all here. It is truly game changing for someone like me coming into it. I am already having a blast with it. 

The basic design is also really well written - for example, Behaviors are really fun and easy ways to code rather than inheriting all the time to make new classes. It is also easy to create new systems that interact with the base native code.

**Problems With MAUI**

If I am understanding correctly, there seem to be an avalanche of bugs and not adequate team members to resolve them. This seems to be an underlying problem that has been going on for some time from what I see from people who were here before me.

For example, I read comments like this: https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-rc-1/#comment-18844 and this https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-rc-1/#comment-18944 with terror.

That user is upset because it sounds like he has been trying to transfer from Xamarin to using MAUI since 2022 (now forced because Xamarin is phasing out) and still has severe bug reports from 2022 going unfixed. This is not unique. The entire comment thread is people upset with bugs present >1 year and not enough resources from Microsoft to fix them. Then the poor Microsoft dev writing the blog trying to answer them when it is obviously not his fault.

To quote that one user linked:

>Currently, MAUI seems to be hovering at the same level as any other random open-source project where features get implemented or fixed based on whichever contributor has the time and/or interest to address it, rather than a flagship UI framework built by the world’s biggest software company. 

**Severity of Existing Bugs**

In just the past 3 weeks I have been trying to replicate core basic necessary functions in MAUI that I would need for any app, I have encountered and submitted the following Bug Reports:

- **Image causes a catastrophic glitch in Android after 1-5 loads of any embedded image in the project** (I solved this for myself by re-writing a new image loading method, but how will others survive this?): https://github.com/dotnet/maui/issues/17857

- **Android ignores requests for heights/widths of AbsoluteLayout/VerticalStackLayout** (which are core objects everyone must use): https://github.com/dotnet/maui/issues/17883

- **AbsoluteLayout/VerticalStackLayout render 1 pixel padding around them causing objects to overlap and show through around the edges of the screen on all platforms** (again, core classes everyone needs): https://github.com/dotnet/maui/issues/17882

- **Entire words are going missing from Label display in Android under clearly reproducible conditions** (how can anyone build an app where entire words may disappear in the biggest mobile OS on earth?): https://github.com/dotnet/maui/issues/17884

- **Shadow has wrong internal pixel scaling in Android** (likely a 1-2 hour at most fix to apply, as I already found the issue and gave a quickfix): https://github.com/dotnet/maui/issues/17886

- **Editor in iOS is completely broken in .NET 7 (doesn't scroll at all) and in .NET 8 exhibits bizarre unusable function in its scrolling, going entirely off the screen if you keep typing** (how can you build an app with any significant user input for iPhone users without working Editor function?): https://github.com/dotnet/maui/issues/17757

- **Border is truncating the top left corner of anything nested into it** (Border is meant to be our primary 'drawn element' on screen - it is the primary display workhorse of MAUI along with the Layout elements & Label): https://github.com/dotnet/maui/issues/17761

- **MediaElement cannot load any embedded videos at all** (this is "Community Toolkit" but it is also the suggested video player for the project, and it is totally unusable as it can't even load videos): https://github.com/CommunityToolkit/Maui/issues/1433

- **The AppWindow.TitleBar in Windows is completely broken, not allowing you to set its text or get its height, and the AppWindow.Position function associated with it is broken too**: https://github.com/dotnet/maui/issues/17679

- **Shadows on Labels disappear in Windows and don't return when Label opacity is tweened between 0 and 1**: https://github.com/dotnet/maui/issues/17916

**The Big Issue**

What is disturbing to me is none of these are high level functions. They are all also possible to reproduce with 40-180 lines of code. They are not using any complex systems. They are basic things like Layout, Label, Shadow, and Border. I did not go looking for any of them either. I just started trying to build very basic displays to test simple functions I need and issue after issue came up causing bizarre problems I had to track down.

I post the list of what I have seen in just the past 3 weeks to illustrate: What do you see when you read this list? Are these esoteric or complex reproduction conditions? 

They are not. They are very core and basic functions. With the exception of the last few perhaps (which some people could survive without), all will potentially cripple anyone developing any app no matter who is building it or what their design is at all. They are all using just the most basic functions of MAUI that we all will need like Border/Layout/Editor/Image.

How can anyone build an effective app when these (the most basic fundamental tools like Border, Editor, Image, and Layout) are not working correctly in simple easily reproduced conditions?

I presume perhaps looking optimistically no one simply highlighted these issues in a clear way. But I also see the loads of people complaining that MAUI has had major reported bugs since 2022 that remain unfixed and that Microsoft is not allocating resources to fix them, and it makes me genuinely concerned in the bigger sense. More and more issues are posted each day but no one is assigned to fix them and the list keeps growing. I only see 2-3 guys on this page from Microsoft replying or looking at things.

**Questions**
- Is this truly a serious project for Microsoft? I sincerely hope it is because it fills a major need in the market at least from a developer perspective. Microsoft has a market cap of _$2.4 trillion_ with $72 billion profit last year. It just seems if they wanted to make something work they certainly are not shy on money to do it.
- How many people are working on the project currently? Paid or volunteer?
- Is Microsoft or the team planning to ramp up focus and resources in anticipation of the May 2024 discontinuation of Xamarin Forms?
- Can we expect any longevity out of MAUI going forward for maintenance?

I am sincerely not writing this to troll or abuse or put down anyone involved in the project or at Microsoft. I am incredibly grateful for the project as I need it very much. I already love the system. When I found the Image loading bug above, I was able to solve it in just a few hours by writing my own image loading code because the basic system is highly versatile. This is a vote of confidence to how well designed it generally is. It is a far better design than anything else on the market even in its damaged state.

Perhaps people in the trenches at Microsoft are not in a position to comment. But is there anyone that can comment on the seriousness that MAUI is being developed with? Maybe the people up high don't realize what's happening in the trenches? Is there any reasonable plan for correcting the >2000 reported active bugs? Or at least getting the basic core functions working like Border/Layout/Label/Editor to a production level? If so, on what timeline?

Thanks for any thoughts and all your continued efforts.

### Public API Changes

```
public void MicrosoftFixMauiPlease(int acceptableIssueCount) {

    while (issueList.Count() > acceptableIssueCount) {
        Worker newWorker = Microsoft.HR.HireWorker();
        newWorker.AssignToMAUI(issueList[0]);
        issueList.RemoveAt(0);
    }
}
```

### Intended Use-Case

Have a beautiful working MAUI ecosystem everyone can enjoy for years to come, replacing all the other garbage multi-platform systems. :)

**Comment by akhanalcs** [comment]
I love MAUI and want it to succeed! I hope leadership at Microsoft sees potential in it and invest in resources accordingly.

PS: Love your `MicrosoftFixMauiPlease` method btw. 😃

**Comment by Vroomer** [comment]
So far Microsoft doesn't show any signs of taking the community seriously and being fully committed to this project. As far as I know they only hired couple more PMs, so that every MAUI developer can have a personal PM I guess. Promises of Microsoft dogfeeding MAUI seem to be mostly empty words. Apparently nobody in Microsoft wants to touch such an unstable platform, so we are supposed to be the ones to take the burden for them. @davidortinau chooses to hide away from these topics and doesn't provide any meaningful assurances about the future. That alone speaks volumes.

It wouldn't be such a problem, if we weren't lied to about MAUI. MAUI was in reality nowhere near GA when they first published it and many others are now forced to switch from Xamarin as it's at the end of the life. The community is rightfuly frustrated and angry.

From my experience, the difference between .NET 7 and .NET 8 is not a significant step forward, especially on Windows. The amount of regression we experienced during .NET 8 previews is appalling so the whole development feels like 3 steps forward, 2 steps back. Furthermore the yearly release cadence tied to .NET versions will once again leave everyone without considerable update for at least half a year, because most bugfixes are not ported to older version and with all the regression you can't expect first previews of .NET 9 to be usable. Waiting this long for further bug fixes is unacceptable.

Yes, MAUI can be amazing product. The ability to go native at will is great advantage as well as #C and .NET ecosystem. Things however have to change if Microsoft wants to keep any trust of the community.

On the bright side, at least we have the memes: https://twitter.com/dotnetmeme

PS: I would like to thank to those few developers that do what they can to improve MAUI. This is in no way targeted against you.

**Reply by jonmdev** [reply]
Holy shit @Vroomer, wait a second ... Fixes aren't ported back into prior versions? That is crazy. Why on earth?

So then what I see here: "The release date for .NET 8.0 is expected to be in November 2023" and this confirms they are annual releases as you said:

https://www.c-sharpcorner.com/article/future-of-dot-net/

In other words if all these bugs I listed (or any others - those are just the ones I personally hit the wall against) are not fixed by NOVEMBER (_1 more month_), we are stuck with them for ANOTHER YEAR???

ie. No working basic Editor function in iOS? No working Border nesting? Layouts not rendering properly or listening to height/width? No working embedded videos? Numerous Shadow bugs? Entire words going missing on Android Labels? Image loading from resources using default method breaking apps on Android?

We would be stuck with all these things until November, 2024? That basically means no one can release anything until November 2024 or as you said we will have to use .NET 9.0 experimentally for our releases while waiting for basic things to be fixed. Which is I guess where we will end up unless something is done.

I know the concept of "crunch" gets a bad reputation in the industry, but is there any way Microsoft could allocate some extra resources for the next month to at least fix the major bugs so we have a working .NET 8.0 version of MAUI we can actually release apps with?

Could anyone with any sway at Microsoft make an appeal to the higher ups to get a bit of extra help?

Currently I don't know how anyone could release something. Just the first 7 out of 10 bugs I listed will cripple virtually any app on iOS/Android (the two OS's virtually every phone uses, and most of us are probably developing for mobile why we are using this). eg. Not having a working iOS Editor makes releasing anything that requires user text input completely impossible on iOS (ie. the entire mobile iPhone market).

Otherwise when May 2024 comes after that everyone is going to be treading water without a raft in the middle of the ocean. If long time Xamarin developers are frustrated now I can only imagine how they will feel then.

I think it is important as you say to recognize the distinction between the developers working on the project and the people allocating resources to make it work. It doesn't matter how good of a developer you are - there are only so many hours in the day. If there are not enough people working on the project, they will be set up for failure and it is not fair to anyone.

I don't know how the management structure at Microsoft works or how hierarchical or flexible they are. But I hope that it is at least an open enough environment that if there is a looming catastrophe (which is what this is looking like for .NET MAUI/Xamarin) someone can at least do something responsive to fix it before it happens.

From where I am looking, it just appears like they need a few more workers to help crunch it a bit for .NET 8.0 and get some basic necessities working ASAP in advance of the known obsolescence of Xamarin in May 2024. If at least the basic classes in .NET 8.0 MAUI are working for their core functions (Border/Layout/Editor/Label/Image/Shadow) then at least no one can say "you can't make an app in MAUI."

But if we have .NET 8.0 released, these basic features aren't working, AND Xamarin has been shot in the back of the head, how on earth are people going to react? Once you lose the community, you lose the product, and that triggers a death spiral. I don't want to see that. I just want to see this working and healthy so I can use it and keep using it going forward.

**Reply by Vroomer** [reply]
After GA of .NET 8, you can expect about 3-4 months of service releases, before the development on .NET 9 starts. After that the attention is gradually shifting towards new .NET version as many bug fixes require underlying breaking changes and they don't want to introduce them into the stable version. There were also multiple cases of bug fixes ported back into .NET 7 that caused terrible regressions. Ever since that they have been very cautious about which bugfixes are backported and only minority of them made it to the .NET 7. So there will still be service releases for .NET 8, but without most of the bug fixes. At least that is how it's been this year as you can see it here: https://github.com/dotnet/maui/releases

**Comment by jonmdev** [comment]
I just have to point out something else. I looked around the Backlog section (which is where all my reported issues are now moved). It is **1600 issues long**: https://github.com/dotnet/maui/milestone/14

And the first issue on the list I saw was this one: https://github.com/dotnet/maui/issues/10401 from Sept 29, 2022: **"Shadow.Radius on Android13 [API 33] is broken."**

There are 21 comments of people complaining and begging for a fix for now 13 months. The same issue was linked at least two other places: https://github.com/dotnet/maui/issues/16440 & https://github.com/dotnet/maui/issues/16311

It is the same Android Shadow scaling issue I posted in my OP with a fix: https://github.com/dotnet/maui/issues/17886

This issue had been sitting for **OVER ONE YEAR** with at least three threads and no solution. It took me **less than one hour** to solve it, and I am not some superstar coder pro genius. I had just happened to see the `FromPixels()` and inverse `ToPixels()` code MAUI uses for scaling Android in other parts of the underlying code a few days prior so it was obvious this had to then be the likely issue.

But this should have been equally obvious to any Microsoft employee working on it too. `ToPixels` and `FromPixels` are all over the Android MAUI code. Search `ToPixels` and you will find it used 89 places to scale things and `FromPixels` 48 times:

https://github.com/search?q=repo%3Adotnet%2Fmaui%20topixels&type=code
https://github.com/search?q=repo%3Adotnet%2Fmaui+frompixels&type=code

These are not some mystery functions in MAUI. They are the main way Android scales to MAUI and back. Almost every parameter is scaled with this in Android. It is simply beyond plausible that if there was a scaling issue with _anything_ in Android the same people who typed `ToPixels` 89 times and `FromPixels` 48 times in the code would have no idea that was the reason. That's just not possible. The people who wrote it over and over would have to know. 

Thus the only logical explanations I can think of then are (i) there is simply no one (or nearly no one) there to even look at the issues or solve them, and/or (ii) the people who wrote MAUI are not even on MAUI anymore. The issues are just being dumped into the void and abandoned. 

Even the ones that should be a slam dunk are being abandoned which has to mean either there is no one there to fix them or the people there trying to fix them aren't the same ones who wrote the system and don't know the system well enough to know how to fix them. 

1600 issues in backlog and most of mine just went too.

☹️

I am suspecting the people who actually wrote MAUI are long gone. Tell me I'm interpreting this wrong.

**Reply by jonmdev** [reply]
Okay well I partly take it back. It looks like the Android Shadow issue in that thread was a specific bug in the Android OS unrelated to the scaling issue I reported. 

So perhaps I can go back to my charitable conclusion from the OP that it is just no one had reported these issues before. Based on that, I decided I will bite the bullet and stick with MAUI. 

As I see it, iOS/Android are mature OS systems that are not rapidly changing. I only need the basic core MAUI functions as follows: 

Editor, Border, Image, Layout (Vertical/Horizontal/Stack/Flex), MediaElement and that is it. I don't use xaml or anything else. I write my own scroll views, click functions, collection view type objects.

Worse case scenario, I can fork it and maintain it on my own. It is still better than anything else. Once those basic functions are working (and the bugs I listed are the only ones I can find in them so far) then I'm at least stable from there.

But I still agree with the guy I quoted in the OP - this feels more like abandonware or random open source project than a truly serious Microsoft project. Which is a bit disappointing. When I first started considering MAUI I thought "Oh amazing, it's Microsoft, well surely if there are any problems things will get fixed. No stress."

We'll see.

**Comment by Strypper** [comment]
Tried the new RC1, really really disappointed. Guess this is my last draw on this dumb framework

**Comment by zfcflower** [comment]
As a developer, we certainly want the product to get better and better, but rebuilding developers' confidence in MAUI is hard. It seems to me that MS spent $5 billion and just didn't want it to die so quickly. The MAUI team needs to take a look at their place in the MS product line

**Comment by ne0rrmatrix** [comment]
I am just an amateur dev. I have been working on trying to get some of the media element bugs fixed. With dotnet 8.x now released it has resulted in situations where tools like the community toolkit for maui has from my point of view a ton of new bugs. Some of which are not in any way code from the toolkit. An example is the menu's used in Media Element for displaying the options to choose source is just not working on Mac Catalyst. It is an ongoing issue with Microsoft.Maui.Controls. It has been in fact fixed. Then it has had regression.

Another example of a toolkit issue is: [https://github.com/CommunityToolkit/Maui/issues/1433](https://github.com/CommunityToolkit/Maui/issues/1433) which I have a PR for atm. It is being looked at and if approved should be out soon. I have another fix for HLS crash when invalid source is selected in Windows. It resulted in CTD for Maui community toolkit sample. That is also an active PR that is awaiting approval. 

I am just one of many people from the community working on fixes. These fixes are in fact only my second time I have tried to give back to the community. I am still learning and have a long way to go. I have found and not had a chance to report other bugs in Maui. I just spend most of my time working on the ones that I can fix easily. The ones that baffle me I leave to others. 

An example there is an issue where if you select a new source for MediaElement playback position tracking does not work in android and IOS. But it works fine in windows. Tracking down what exactly goes wrong is hard. I have spent two days on it and I am moving on. I am too new to figure that one out.

**Comment by streetwiseherc** [comment]
Microsoft has spent $13 billion on Open AI in hopes that one day ChatGPT will be able to fix those 2,500 bugs in MAUI.

![image](https://github.com/dotnet/maui/assets/4156126/7b50f79c-87bd-406a-8e5f-127179e3f1b6)

**Reply by Strypper** [reply]
not to mention, the fixes they announced are available but the users end up having other bugs while do the migration from 7 to 8 
and they give 0 fvk about it

**Comment by AlleSchonWeg** [comment]
The MAUI death will coming. Sooner or later. MS has mada a couple of mistakes in my opinion:
- Force users to migrate from XF to MAUI. Migration is complicated, takes a lot of time and it's not safe that the App behaves the same.
- They had to focus on iOS and Android. Not sure if somebody realy needs MAUI for windows or mac.
- Retirement of VS for mac. We must use VS-Code (not working with XF) to code and publish an app to the app store. 
- Dev experience with VS is sometimes crazy.  Build, Clean, Restart VS. Delete bin/obj and repeat. 
- MAUI as a workload. Nuget was easy and the dev had the full version control.
- XF 3rd party librarys are not compatible with MAUI. Many authors not migrate their librarys to MAUI. This takes also a lot of time to find alternatives.
- Bindings are not up to date or abandoned. For example the Google API's for iOS.

I tried to migrate our XF app but it failed. For me the question is: 
Makes it sense to invest weeks or month of time/money to migrate to MAUI. Or should we migrate to flutter/react.

We like c#, thats the reason we use XF. Perhaps MS should make c# as an alternative to dart to write flutter apps.

**Reply by tjiakwokyung28** [reply]
no, dot net maui is the fantastic technology, maui is the best in the future. maui is best and speedest than xamarin, fluter, java etc. i'am development using maui.

**Comment by Atulin** [comment]
MAUI will be a trustworthy UI framework when Microsoft actually starts using it themselves. So far, the direction has been the opposite: all their native apps are migrating to React + WebView. That tells me all I needed to know.

**Comment by Xyncgas** [comment]
Unified UI and native interop > native UI

And instead of using XMAL as the 'device native html' the way it is now, either let me manipulate XAML in .NET like I can in javascript to manipulate html, or build UI completely without and markup and just code only | use blazor syntax for MAUI | compile UI to GPU shader so it can run on any system

We can always run platform native code inside unified UI via dependency injection

**Comment by NCLnclNCL** [comment]
Use avalonia
